### ISSUE TITLE
Webservice issue

### ISSUE THREAD

**Issue by benzon** [thirtybees/thirtybees#434]
Hi,

We are currently trying to integrate our ERP with our TB webshop, but we have run in to a error/bug in TB.

The response of the webservice api shows that there might be a flaw where i tryes to call a multishop function even tho the shop has never bin setup in multishop format.

The company that handles the erp integration send me the following information, i have since verifyed it my self, and it's even a issue in a 100% clean TB install.

https://domain.com/api/combinations/2

The reply

```xml
<?xml version="1.0" encoding="UTF-8"?>
<thirtybees xmlns:xlink="http://www.w3.org/1999/xlink">
    <errors>
        <error>
            <code>
                <![CDATA[0]]>
            </code>
            <message>
                <![CDATA[Wrong argument ($type) in Shop::getSharedShops() method]]>
            </message>
        </error>
    </errors>
</thirtybees>
```

Expectet reply would be

```xml
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
    <combination>
        <id>
            <![CDATA[2]]>
        </id>
        <id_product xlink:href="https://www.domain.dk/api/products/3">
            <![CDATA[3]]>
        </id_product>
        <location></location>
        <ean13>
            <![CDATA[4024144511631]]>
        </ean13>
        <upc></upc>
        <quantity>
            <![CDATA[10]]>
        </quantity>
        <reference>
            <![CDATA[O511641-1]]>
        </reference>
        <supplier_reference></supplier_reference>
        <wholesale_price>
            <![CDATA[0.000000]]>
        </wholesale_price>
        <price>
            <![CDATA[0.000000]]>
        </price>
        <ecotax>
            <![CDATA[0.000000]]>
        </ecotax>
        <weight>
            <![CDATA[0.000000]]>
        </weight>
        <unit_price_impact>
            <![CDATA[0.000000]]>
        </unit_price_impact>
        <minimal_quantity>
            <![CDATA[1]]>
        </minimal_quantity>
        <default_on></default_on>
        <available_date>
            <![CDATA[0000-00-00]]>
        </available_date>
        <associations>
            <product_option_values nodeType="product_option_value" api="product_option_values">
                <product_option_value xlink:href="https://www.domain.dk/api/product_option_values/26">
                    <id>
                        <![CDATA[26]]>
                    </id>
                </product_option_value>
            </product_option_values>
            <images nodeType="image" api="images/products">
                <image xlink:href="https://www.domain.dk/api/images/products/3/8">
                    <id>
                        <![CDATA[8]]>
                    </id>
                </image>
            </images>
        </associations>
    </combination>
</prestashop>
```

Hope this is a issue that will be looked at quick since we cant rly go in to the air with our b2b shop before this is working again ;(

**Comment by benzon** [comment]
After some digging the issue is locatet to classes/Combination.php

If PS version is uploaded insted the respons is correct again, now it's just finding the issue in TB version of that file.

**Comment by soulilya** [comment]
I also have an issue when i try to add product via webservice - get 500 Server Error. Webserver log is empty!

**Comment by benzon** [comment]
Not sure if the fix i just did a pull request on will fix that.

**Comment by benzon** [comment]
@soulilya what api call do you do that causes the 500 server Error?

Then it's easyer to track down if you give an example on what you do, and how you do it :)

**Comment by soulilya** [comment]
For clear, when i got an issue to make sure that i all doing right:
1. i take product xml response 
2. do calls and delete all tags, that api ask for
3. 500 Server Error
Like this example i successfully import manufacturers, categories, features, features values.
P.S. TB version 1.0.3

**Comment by firstred** [comment]
@soulilya Have you tried this change we just discovered? https://github.com/thirtybees/thirtybees/pull/436/files

**Comment by benzon** [comment]
Try out the changes that @firstred linked to it might fix the issue.

If you could share the url's you call with content ex.

/api/orders?filter%5Bdate_upd%5D=%3E%5B2018-01-09%5D&date=1&display=full&sort=id_DESC&limit=0%2C100

Then it's possible to test, and easyer to find a fix if there is a issue.

**Comment by soulilya** [comment]
i dont use combinations. Easer to find issue if there is traceback of errors. I set only necessary fields and still get a error. You dont get errors when import products?

**Comment by benzon** [comment]
again from the api call you make bugs can be tracked down to, aint easy to fix a bug when you dont share what it is you are doing when you get a error 500, i can verify if it's an issue in general or it's a issue located in your installation.

**Comment by benzon** [comment]
And is it something that works in PS but is broke in TB? have you tryed to upload PS version of classes/Product.php to your TB site, something like that wil help narrow down the issue.

But being closed up about how you do it dos not solve the issue, no one can find an error out from very little information.

**Comment by benzon** [comment]
@soulilya like i said i would love to dig in to the issue, but i need some more info on what you are doing, to figure out if it's on specific part or what is going on with your issue.

**Comment by soulilya** [comment]
Here my xml:
<thirtybees xmlns:xlink="http://www.w3.org/1999/xlink">
<product>
	<id/>
	<id_manufacturer>7</id_manufacturer>
	<id_supplier>0</id_supplier>
	<id_category_default>8</id_category_default>
	<new>1</new>
	<cache_default_attribute/>
	<id_default_image/>
	<id_default_combination/>
	<id_tax_rules_group/>
	<position_in_category/>
	<type/>
	<id_shop_default/>
	<reference>MOD807-PL-02-60-W</reference>
	<supplier_reference>98351-MOD807-PL-02-60-W</supplier_reference>
	<location/>
	<width>0</width>
	<height>0</height>
	<depth>0</depth>
	<weight>0</weight>
	<quantity_discount/>
	<ean13/>
	<upc/>
	<cache_is_pack/>
	<cache_has_attachments/>
	<is_virtual/>
	<on_sale>0</on_sale>
	<online_only/>
	<ecotax/>
	<minimal_quantity/>
	<price>11710</price>
	<wholesale_price/>
	<unity/>
	<unit_price_ratio/>
	<additional_shipping_cost/>
	<customizable/>
	<text_fields/>
	<uploadable_files/>
	<active>0</active>
	<redirect_type/>
	<id_product_redirected/>
	<available_for_order/>
	<available_date/>
	<condition/>
	<show_price/>
	<indexed/>
	<visibility/>
	<advanced_stock_management/>
	<date_add/>
	<date_upd/>
	<pack_stock_type/>
	<meta_description><language id="1"/></meta_description>
	<meta_keywords><language id="1"/></meta_keywords>
	<meta_title><language id="1"/></meta_title>
	<link_rewrite><language id="1"/></link_rewrite>
	<name><language id="1">MOD807-PL-02-60-W &#1055;&#1086;&#1076;&#1074;&#1077;&#1089;&#1085;&#1086;&#1081; &#1089;&#1074;&#1077;&#1090;&#1086;&#1076;&#1080;&#1086;&#1076;&#1085;&#1099;&#1081; &#1089;&#1074;&#1077;&#1090;&#1080;&#1083;&#1100;&#1085;&#1080;&#1082; Maytoni Nola</language></name>
	<description><language id="1"/></description>
	<description_short><language id="1"/></description_short>
	<available_now><language id="1"/></available_now>
	<available_later><language id="1"/></available_later>
<associations>
<categories>
	<category>
	<id>8</id>
	</category>
</categories>
<images>
	<image>
	<id/>
	</image>
</images>
<combinations>
	<combination>
	<id/>
	</combination>
</combinations>
<product_option_values>
	<product_option_value>
	<id/>
	</product_option_value>
</product_option_values>
<product_features>
	<product_feature>
	<id>156</id>
	<id_feature_value>269375</id_feature_value>
	</product_feature>
<product_feature><id>157</id><id_feature_value>277903</id_feature_value></product_feature><product_feature><id>162</id><id_feature_value>269535</id_feature_value></product_feature><product_feature><id>163</id><id_feature_value>269423</id_feature_value></product_feature><product_feature><id>164</id><id_feature_value>269424</id_feature_value></product_feature><product_feature><id>165</id><id_feature_value>269424</id_feature_value></product_feature><product_feature><id>166</id><id_feature_value>269472</id_feature_value></product_feature><product_feature><id>167</id><id_feature_value>269579</id_feature_value></product_feature><product_feature><id>168</id><id_feature_value>269510</id_feature_value></product_feature><product_feature><id>169</id><id_feature_value>269511</id_feature_value></product_feature><product_feature><id>170</id><id_feature_value>269385</id_feature_value></product_feature><product_feature><id>171</id><id_feature_value>269386</id_feature_value></product_feature><product_feature><id>172</id><id_feature_value>269387</id_feature_value></product_feature><product_feature><id>173</id><id_feature_value>269588</id_feature_value></product_feature><product_feature><id>174</id><id_feature_value>269415</id_feature_value></product_feature><product_feature><id>175</id><id_feature_value>269415</id_feature_value></product_feature><product_feature><id>176</id><id_feature_value>269404</id_feature_value></product_feature><product_feature><id>177</id><id_feature_value>269582</id_feature_value></product_feature><product_feature><id>178</id><id_feature_value>269481</id_feature_value></product_feature><product_feature><id>179</id><id_feature_value>269405</id_feature_value></product_feature><product_feature><id>180</id><id_feature_value>269395</id_feature_value></product_feature><product_feature><id>181</id><id_feature_value>269579</id_feature_value></product_feature><product_feature><id>182</id><id_feature_value>269472</id_feature_value></product_feature><product_feature><id>183</id><id_feature_value>269387</id_feature_value></product_feature><product_feature><id>184</id><id_feature_value>269633</id_feature_value></product_feature><product_feature><id>185</id><id_feature_value>269415</id_feature_value></product_feature><product_feature><id>186</id><id_feature_value>269415</id_feature_value></product_feature></product_features>
<tags>
	<tag>
	<id/>
	</tag>
</tags>
<stock_availables>
	<stock_available>
	<id/>
	<id_product_attribute/>
	</stock_available>
</stock_availables>
<accessories>
	<product>
	<id/>
	</product>
</accessories>
<product_bundle>
	<product>
	<id/>
	<quantity/>
	</product>
</product_bundle>
</associations>
</product>
</thirtybees>

**Comment by soulilya** [comment]
I tryed remove all empty fields, remove asssociations and still nothing.

**Comment by soulilya** [comment]
I figure out that problem appear when i add categories in associations.

**Comment by soulilya** [comment]
> have you tryed to upload PS version of classes/Product.php to your TB site, something like that wil help narrow down the issue.
Still nothing!

**Comment by benzon** [comment]
Hmm might be an issue in the Category.php then if it happens when you add Categorys in associations, but ill have to get back to you on that, just having some stuff i need to sort out first.

**Comment by soulilya** [comment]
No, Product.php

**Comment by soulilya** [comment]
Product.php  from 7070 line nuber:
`                $sqlValues = '';
                $ids = array_map('intval', $ids);
                foreach ($ids as $position => $id) {
                    $sqlValues[] = '('.(int) $id.', '.(int) $this->id.', '.(int) $position.')';
                }`
Init string and than add array value? Really?
Fix please:
`$sqlValues = array();`

And why they code errors? Why hide errors from user? Print all traceback.

**Comment by mke-tb** [comment]
@soulilya I don't know if you solved your issue but the lines you mentioned (`Product.php:7070) were updated to handle the issue better, perhaps you know whether your problem is still reproducible?

**Comment by getdatakick** [comment]
Product category association was fixed in commit ebf6f91131a7adf43a761e39cf51f8587a549a8a. This issue is no longer reproducible. Closing
